src/client.py

```diff
@@ -0,0 +1,362 @@
+        :type method: str
+        :param parameters: (optional) parameters for request.
+        :type parameters: dict: (../?key=val), str: (../?str).
+        :param data: (optional) data. Defaults to (data=None).
+        :return: JSON object containing MERMAID API data.
+        :rtype: dict, None
+        """
+        # Creates full URL path.
+        prep_url = self.url + subdirectory
+        # Prepares Request and send from Client class Session.
+        req = requests.Request(method, url=prep_url, params=parameters, data=data)
+        prepped = self.session.prepare_request(req)
+        resp = self.session.send(prepped)
+
+        # Returns JSON if response code OK.
+        if resp.status_code == requests.codes.ok:
+            return resp.json()
+        else:
+            if resp.status_code == 401:
+                print('api_root response code: ' + str(resp.status_code) + ' -- Attempt token refresh')
+            else:
+                print('api_root response code: ' + str(resp.status_code))
+            return None
+
+    def api_projects(self, name=None, id=None, resource=None, filter=None, filter_val=None):
+        """
+        Helper class function used to return projects path '/projects/<project_id>/' based on name or id of project.
+        See https://mermaid-api.readthedocs.io/en/latest/projects.html#project-resources for more information.
+        :param name: (optional if 'id' provided) MERMAID project name.
+        :type name: str
+        :param id: (optional if 'name' provided) MERMAID project ID.
+        :type: id: str
+        :param filter: MERMAID project resource filters.
+        :type filter: str
+        :param filter_val: (optional) Required for resource filters requiring values. eg. (size_min/size_max,
+        count_min/count_max, length_min/length_max, len_surveyed_min/len_surveyed_max,
+        sample_date_before/sample_date_after).
+        :type filter_val: str, int
+        :param resource: MERMAID project entity resources.
+        :return: projects path, returns None if invalid project name or id.
+        :rtype: str, None
+        """
+        # Generate path.
+        path = None
+        if resource:
+            if id:
+                path = 'projects/' + id + '/' + resource + '/'
+            elif name:
+                p_id = self.get_project_id(name=name)
+                # Nested if instead of elif name and self.get_project_id(name=name) to improve performance by reducing
+                # the number of API calls for name or id NoneType check.
+                if p_id:
+                    path = 'projects/' + p_id + '/' + resource + '/'
+        else:
+            if id:
+                path = 'projects/' + id + '/'
+            elif name:
+                p_id = self.get_project_id(name=name)
+                if p_id:
+                    path = 'projects/' + p_id + '/'
+
+        # API call and filters check.
+        if path:
+            if filter and filter_val:
```

Comment:
Again here my approach would be:

```
payload = None
if filter:
    payload = filter
if filter and filter_val:
    payload = {filter: filter_val}
return self.api_root(path, parameters=payload)
```

You shouldn't have the final default `return path` because that's just a string, not the result of an api call.